Add LangChainClient and notebook demonstrating Running a chain on a dataset

vowelparrot opened the pull request:
Will open a different one with a squash of this later, but draft for testing out:

Further unit tests needed for the client methods. Also would appreciate preferences on things like
- Which directory to place the notebook in
- Is `langchain/client` a good place for the client
- Should I rm a lot of the async methods? (the ones that were wanted are the chain running ones so we can do multiple in parallel, but the others were so that I'd call mostly async methods from the async method, etc. Fine removing for now for many reasons)
- Naming of our CRUD verbs. Currently the API says "Read" for both the single and multiple return actions. I currently have "Read" and "list" in the client. 
- I used an asyncio Queue so we don't have to block on each batch end. There may be some issues I'm not familiar with that would make that less viable at this point.

Think it's probably best to wait to expose on the top level until later also.

vowelparrot committed:
[WIP] Example Notebook running a chain on a dataset

agola11 commented on langchain/callbacks/tracers/langchain.py:
```diff
@@ -138,6 +138,7 @@ def _convert_run(run: Union[LLMRun, ChainRun, ToolRun]) -> Run:
             session_id=run.session_id,
             run_type=run_type,
             parent_run_id=run.parent_uuid,
+            example_id=run.example_id,
```
We don't use this now, but the proper field is `reference_example_id` on the `Run` schema

vowelparrot replied:
Got it so I'll change to `reference_example_id=run.example_id,`

agola11 commented on langchain/callbacks/manager.py:
```diff
@@ -44,11 +44,13 @@ def get_openai_callback() -> Generator[OpenAICallbackHandler, None, None]:
     cb = LangChainTracer()
     session = cb.load_session(session_name)
+    if example_id:
+        cb.example_id
```
cb.example_id = example_id

agola11 commented on docs/use_cases/evaluation/evaluating_traced_examples.ipynb:
let's wrap this all in a client that holds the endpoint, api key, etc. this way we don't have to pass it to all these one-off methods

vowelparrot committed:
Update Client

vowelparrot committed:
add testing

vowelparrot committed:
Add del

vowelparrot committed:
Change Name

vowelparrot committed:
Rerun

vowelparrot committed:
add html repr

vowelparrot committed:
Add tests

vowelparrot committed:
Switch to using an asyncio Queue instead of batches

vowelparrot committed:
Let 'Outputs' be None

vowelparrot committed:
Migrating to new endpoints

vowelparrot committed:
Visual Studio Code/Github Codespaces Dev Containers  (#4035) (#4122)

Having dev containers makes its easier, faster and secure to setup the
dev environment for the repository.

The pull request consists of:

- .devcontainer folder with:
- **devcontainer.json :** (minimal necessary vscode extensions and
settings)
- **docker-compose.yaml :** (could be modified to run necessary services
as per need. Ex vectordbs, databases)
    - **Dockerfile:**(non root with dev tools)
- Changes to README - added the Open in Github Codespaces Badge - added
the Open in dev container Badge

Co-authored-by: Jinto Jose <129657162+jj701@users.noreply.github.com>

vowelparrot committed:
Visual Studio Code/Github Codespaces Dev Containers  (#4035) (#4122)

Having dev containers makes its easier, faster and secure to setup the
dev environment for the repository.

The pull request consists of:

- .devcontainer folder with:
- **devcontainer.json :** (minimal necessary vscode extensions and
settings)
- **docker-compose.yaml :** (could be modified to run necessary services
as per need. Ex vectordbs, databases)
    - **Dockerfile:**(non root with dev tools)
- Changes to README - added the Open in Github Codespaces Badge - added
the Open in dev container Badge

Co-authored-by: Jinto Jose <129657162+jj701@users.noreply.github.com>

vowelparrot committed:
Update V2 Tracer to be compatible with new data model

Add tenant ID's and update UUID's in runs and sessions

vowelparrot committed:
Update V2 Tracer to be compatible with new data model

Add tenant ID's and update UUID's in runs and sessions

vowelparrot committed:
Update session

vowelparrot committed:
Update session

vowelparrot committed:
Update session

vowelparrot committed:
further linting

vowelparrot committed:
further linting

vowelparrot committed:
Merge branch 'vwp/add_tenant' into vwp/eval_examples

vowelparrot committed:
Update uuid typing and the works

vowelparrot committed:
change name

agola11 commented on langchain/client/langchain.py:
```diff
@@ -0,0 +1,310 @@
+        _raise_rich_error(response)
+        return [Example(**dataset) for dataset in response.json()]
+
+    async def alist_examples(
```
would use `list` for getting multiple, `read` for getting one

agola11 commented on langchain/client/langchain.py:
```diff
@@ -0,0 +1,310 @@
+    """Client for interacting with the LangChain+ API."""
+
+    api_key: str = Field(default=None, env="LANGCHAIN_API_KEY")
+    api_url: str = Field(default="http://localhost:8000", env="LANGCHAIN_ENDPOINT")
```
default is fine

vowelparrot committed:
Merge branch 'vwp/add_tenant' into vwp/eval_examples

vowelparrot committed:
Merge branch 'master' into vwp/eval_examples

vowelparrot committed:
add tests

vowelparrot committed:
Update notebook

vowelparrot committed:
Updat with LLM and Chat Models

nfcampos commented on langchain/callbacks/manager.py:
```diff
@@ -66,7 +67,9 @@ def tracing_v2_enabled(
     cb = LangChainTracerV2()
-    session = cb.load_session(session_name)
+    if example_id:
+        cb.example_id = UUID(example_id)
```
Wouldn't this be better as an optional constructor arg to the tracer?

vowelparrot replied:
It would indeed, I believe

nfcampos commented on langchain/callbacks/tracers/schemas.py:
```diff
@@ -115,7 +119,6 @@ class Run(BaseModel):
     parent_run_id: Optional[UUID]
     reference_example_id: Optional[UUID]
     run_type: RunTypeEnum
-    child_runs: List[Run] = Field(default_factory=list)
```
Isnt this still needed for the old tracer?

vowelparrot replied:
I believe the `Run` object is exclusively used in the V2 tracer right now

vowelparrot replied:
Oops I was wrong it's used by the n ew one as well still

nfcampos commented on langchain/client/langchain.py:
```diff
@@ -0,0 +1,521 @@
+        num_repetitions: int = 1,
+        session_name: Optional[str] = None,
+        verbose: bool = False,
+    ) -> Dict[str, Any]:
```
A suggestion here: I think what you're trying to achieve here is a  version of `asyncio.gather` with a concurrency limit, which can be implemented this way https://stackoverflow.com/a/61478547

vowelparrot replied:
Oh cool love it .So make N semaphors then and grab?

vowelparrot replied:
Will try to apply

nfcampos commented on langchain/client/langchain.py:
```diff
@@ -0,0 +1,521 @@
+                )
+                if verbose:
+                    print(f"{i+1} processed", flush=True, end="\r")
+                if i > 5:
```
What does 5 mean?

vowelparrot replied:
oops sorry was debugging. I'll turn this into a draft again

nfcampos reviewed (commented):
Some comments

vowelparrot committed:
Merge branch 'master' into vwp/eval_examples

vowelparrot committed:
Update V2 Tracer

- Update the RunCreate object to work with recent changes
- Adjust default persist_session behavior to attempt to
load the session if it exists
- Raise more useful HTTP errors for logging
- Add unit testing

vowelparrot committed:
Update V2 Tracer

- Update the RunCreate object to work with recent changes
- Adjust default persist_session behavior to attempt to
load the session if it exists
- Raise more useful HTTP errors for logging
- Add unit testing

vowelparrot committed:
Update V2 Tracer

- Update the RunCreate object to work with recent changes
- Adjust default persist_session behavior to attempt to
load the session if it exists
- Raise more useful HTTP errors for logging
- Add unit testing

vowelparrot committed:
Merge branch 'vwp/lcp_client' into vwp/eval_examples

vowelparrot committed:
Merge branch 'vwp/lcp_client' into vwp/eval_examples

vowelparrot committed:
Merge branch 'vwp/lcp_client' into vwp/eval_examples

vowelparrot commented:
Incorporated feedback into https://github.com/hwchase17/langchain/pull/4198
